src/Data/Semigroup/Linear.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE StandaloneDeriving #-}
+
+-- | = The linear semigroup hierarchy
+--
+-- TODO: documentation
+
+module Data.Semigroup.Linear
+  ( Semigroup(..)
+  , Monoid(..)
+  , LEndo(..), appLEndo
+  , Top, throw
+  , module Data.Semigroup
+  )
+  where
+
+import Prelude.Linear.Internal.Simple
+import Data.Semigroup hiding (Semigroup(..))
+import qualified Data.Semigroup as Prelude
+import qualified Prelude
+import qualified Unsafe.Linear as Unsafe
+
+class Prelude.Semigroup a => Semigroup a where
+  (<>) :: a ->. a ->. a
+
+class (Semigroup a, Prelude.Monoid a) => Monoid a where
+  {-# MINIMAL #-}
+  mempty :: a
+  mempty = mempty
+  -- convenience redefine, so that nonlinear Data.Semigroup doesn't have
+  -- to be imported to use mempty
+
+---------------
+-- Instances --
+---------------
+
+instance Semigroup () where
+  () <> () = ()
+
+-- | The type of linear endomorphisms.
+newtype LEndo a = LEndo (a ->. a)
+
+-- TODO: have this as a newtype deconstructor once the right type can be
+-- correctly inferred
+appLEndo :: LEndo a ->. a ->. a
+appLEndo (LEndo f) = f
+
+instance Prelude.Semigroup (LEndo a) where
+  LEndo f <> LEndo g = LEndo (f . g)
+instance Prelude.Monoid (LEndo a) where
+  mempty = LEndo id
+instance Semigroup (LEndo a) where
+  LEndo f <> LEndo g = LEndo (f . g)
+instance Monoid (LEndo a) where
+
+instance (Semigroup a, Semigroup b) => Semigroup (a,b) where
+  (a,x) <> (b,y) = (a <> b, x <> y)
+instance (Monoid a, Monoid b) => Monoid (a,b)
+
+instance Semigroup a => Semigroup (Dual a) where
+  Dual x <> Dual y = Dual (y <> x)
+instance Monoid a => Monoid (Dual a)
+
+newtype LWrap a = LWrap a
+  deriving (Prelude.Semigroup, Prelude.Monoid)
+
+-- This instance is unsafe: do not export LWrap so it cannot be used.
+instance Prelude.Semigroup a => Semigroup (LWrap a) where
+  LWrap a <> LWrap b = LWrap (Unsafe.toLinear2 (Prelude.<>) a b)
```

Comment:
It's actually unsafe for `movable` things too. Unless `(<>)` is hyperstrict in both arguments. Which it rarely is. You really need to force all the linear thunks that you may encounter. Sorry I'd missed that on my first review.

It's totally ok to define these instances manually. They are not big.

Reply:
Ahh this is frustrating - I'll edit. Could you explain why this would fail for movable things?

Reply:
Suppose you have a thunk of type `True`. It may have, in its closure, some linear variables. As long as you haven't pattern-matched on the thunk, then this linear variable hasn't been consumed. Which breaks linearity.

Concretely, let's take boolean disjunction

```haskell
or :: Bool -> Bool -> Bool
or True _ = True
or False b = b
```

I claim that `or`, thus defined, is not a linear function. For it to be a linear function, it would need, per the defintion, that consuming `or x y` exactly once would consume `x` and `y` exactly once. But if `x` is `True`, then consuming `or x y` would not consume `y`.

Booleans are movable alright, but moving them _does_ require deep copy (which is not too deep in this case, but still).

Reply:
I see, thanks!